File: frontend/test/features/task/domain/due_grouping_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smart_todo_app/features/task/domain/entities/task.dart';
import 'package:smart_todo_app/features/task/domain/services/due_grouping.dart';

Task _task(String id, {DateTime? due, TaskStatus status = TaskStatus.todo}) =>
    Task(
      id: id,
      title: id,
      status: status,
      priority: TaskPriority.medium,
      dueDate: due,
      createdAt: DateTime.utc(2026, 7, 1),
      updatedAt: DateTime.utc(2026, 7, 1),
    );

void main() {
  // Saturday 2026-07-18 at 15:00 local time.
  final now = DateTime(2026, 7, 18, 15);

  List<DueSection> group(List<Task> tasks) => groupByDue(tasks, now);

  test('a past-due open task is overdue', () {
    final sections = group([_task('a', due: DateTime(2026, 7, 17, 23, 59))]);

    expect(sections.single.group, DueGroup.overdue);
  });

  test('a task due later today is Oggi, even at 23:59', () {
    final sections = group([_task('a', due: DateTime(2026, 7, 18, 23, 59))]);

    expect(sections.single.group, DueGroup.today);
  });

  test('a task due earlier today but past the hour is overdue', () {
    final sections = group([_task('a', due: DateTime(2026, 7, 18, 9))]);

    expect(sections.single.group, DueGroup.overdue);
  });

  test('tomorrow is Domani', () {
    final sections = group([_task('a', due: DateTime(2026, 7, 19, 8))]);

    expect(sections.single.group, DueGroup.tomorrow);
  });

  test('within the next six days is Questa settimana', () {
    final sections = group([_task('a', due: DateTime(2026, 7, 24))]);

    expect(sections.single.group, DueGroup.thisWeek);
  });

  test('beyond a week is Più avanti', () {
    final sections = group([_task('a', due: DateTime(2026, 7, 25))]);

    expect(sections.single.group, DueGroup.later);
  });

  test('no due date is Senza scadenza', () {
    final sections = group([_task('a')]);

    expect(sections.single.group, DueGroup.noDate);
  });

  test('a DONE task is never overdue: all done tasks group under Completati',
      () {
    final sections = group([
      _task('past-done', due: DateTime(2026, 7, 10), status: TaskStatus.done),
      _task('today-done', due: DateTime(2026, 7, 18, 23), status: TaskStatus.done),
    ]);

    expect(sections.single.group, DueGroup.completed);
    expect(sections.single.tasks, hasLength(2));
  });

  test('sections come out in fixed order with empty groups omitted', () {
    final sections = group([
      _task('later', due: DateTime(2026, 9, 1)),
      _task('overdue', due: DateTime(2026, 7, 1)),
      _task('none'),
      _task('done', status: TaskStatus.done),
      _task('today', due: DateTime(2026, 7, 18, 20)),
    ]);

    expect(sections.map((s) => s.group), [
      DueGroup.overdue,
      DueGroup.today,
      DueGroup.later,
      DueGroup.noDate,
      DueGroup.completed,
    ]);
  });

  test('isOverdue matches the grouping rule', () {
    expect(isOverdue(_task('a', due: DateTime(2026, 7, 17)), now), isTrue);
    expect(
        isOverdue(_task('a', due: DateTime(2026, 7, 17), status: TaskStatus.done), now),
        isFalse);
    expect(isOverdue(_task('a', due: DateTime(2026, 7, 19)), now), isFalse);
    expect(isOverdue(_task('a'), now), isFalse);
  });
}

File: webapp/src/features/tasks/due-grouping.test.ts
import { describe, expect, it } from 'vitest'
import type { Task, TaskStatus } from '@/api/types'
import { dueGroupLabels, groupByDue, isOverdue, type DueGroup } from '@/features/tasks/due-grouping'

/** Mercoledì 15 luglio 2026, ore 12:00 locali. */
const now = new Date(2026, 6, 15, 12, 0, 0)

function task(overrides: Partial<Task> & { id: string }): Task {
  return {
    title: 'Un task',
    status: 'TODO',
    priority: 'MEDIUM',
    tags: [],
    createdAt: '2026-07-01T08:00:00Z',
    updatedAt: '2026-07-01T08:00:00Z',
    ...overrides,
  }
}

/** Data locale con offset in giorni e ora scelta, serializzata come manda il backend. */
function dueIn(days: number, hour = 12): string {
  const date = new Date(2026, 6, 15 + days, hour, 0, 0)
  return date.toISOString()
}

function groupOf(overrides: Partial<Task>): DueGroup | undefined {
  const sections = groupByDue([task({ id: 't1', ...overrides })], now)
  return sections[0]?.group
}

describe('groupByDue', () => {
  it('mette in "In ritardo" ciò che è scaduto e non è fatto', () => {
    expect(groupOf({ dueDate: dueIn(-1) })).toBe('overdue')
    // Stessa giornata di oggi, ma un'ora già passata: è comunque in ritardo.
    expect(groupOf({ dueDate: dueIn(0, 9) })).toBe('overdue')
  })

  it('distingue oggi, domani e questa settimana', () => {
    expect(groupOf({ dueDate: dueIn(0, 18) })).toBe('today')
    expect(groupOf({ dueDate: dueIn(1) })).toBe('tomorrow')
    expect(groupOf({ dueDate: dueIn(2) })).toBe('thisWeek')
    expect(groupOf({ dueDate: dueIn(6) })).toBe('thisWeek')
  })

  it('oltre i sei giorni è "Più avanti"', () => {
    expect(groupOf({ dueDate: dueIn(7) })).toBe('later')
    expect(groupOf({ dueDate: dueIn(365) })).toBe('later')
  })

  it('senza scadenza finisce in "Senza scadenza"', () => {
    expect(groupOf({})).toBe('noDate')
  })

  it('i completati collassano tutti insieme, qualunque sia la scadenza', () => {
    const statuses: TaskStatus[] = ['DONE']
    for (const status of statuses) {
      expect(groupOf({ status, dueDate: dueIn(-10) })).toBe('completed')
      expect(groupOf({ status, dueDate: dueIn(0) })).toBe('completed')
      expect(groupOf({ status })).toBe('completed')
    }
  })

  it('un task in corso e scaduto resta in ritardo', () => {
    expect(groupOf({ status: 'IN_PROGRESS', dueDate: dueIn(-2) })).toBe('overdue')
  })

  it('restituisce le sezioni nell ordine di urgenza, saltando quelle vuote', () => {
    const sections = groupByDue(
      [
        task({ id: 'senza-data' }),
        task({ id: 'fatto', status: 'DONE' }),
        task({ id: 'domani', dueDate: dueIn(1) }),
        task({ id: 'ritardo', dueDate: dueIn(-3) }),
        task({ id: 'oggi', dueDate: dueIn(0, 20) }),
      ],
      now,
    )

    expect(sections.map((section) => section.group)).toEqual([
      'overdue',
      'today',
      'tomorrow',
      'noDate',
      'completed',
    ])
    expect(sections.map((section) => section.label)).toEqual([
      'In ritardo',
      'Oggi',
      'Domani',
      'Senza scadenza',
      'Completati',
    ])
  })

  it('conserva l ordine dei task dentro ogni sezione', () => {
    const sections = groupByDue(
      [
        task({ id: 'primo', dueDate: dueIn(3) }),
        task({ id: 'secondo', dueDate: dueIn(4) }),
        task({ id: 'terzo', dueDate: dueIn(2) }),
      ],
      now,
    )

    expect(sections[0]?.tasks.map((t) => t.id)).toEqual(['primo', 'secondo', 'terzo'])
  })

  it('senza task non produce sezioni', () => {
    expect(groupByDue([], now)).toEqual([])
  })

  it('ha un etichetta per ogni gruppo', () => {
    expect(Object.values(dueGroupLabels)).toHaveLength(7)
  })
})

describe('isOverdue', () => {
  it('è in ritardo solo ciò che ha una scadenza passata e non è completato', () => {
    expect(isOverdue(task({ id: 't', dueDate: dueIn(-1) }), now)).toBe(true)
    expect(isOverdue(task({ id: 't', dueDate: dueIn(1) }), now)).toBe(false)
    expect(isOverdue(task({ id: 't' }), now)).toBe(false)
    expect(isOverdue(task({ id: 't', status: 'DONE', dueDate: dueIn(-1) }), now)).toBe(false)
  })
})
